_labs/lab02.markdown

```diff
@@ -0,0 +1,352 @@
+You can use the `variable` function from `Utils.h` to get the name of an operand from its corresponding LLVM Value.
+
+**Dynamic Analysis**.
+It involves inspecting a running program for information about its state and behavior during runtime; this is in contrast to static analysis which analyzes the properties of code independent of any execution.
+One way to inspect the runtime behavior of a program is by injecting code into the program during compile time; this technique falls under the umbrella term [instrumentation][instrumentation-def].
+For both static analysis in `src/StaticAnalysisPass.cpp` we will have a corresponding dynamic analysis instrumentation in `src/DynamicAnalysisPass.cpp`.
+We have provided you with an implementation for the first analysis which injects a call to `__coverage__` function before every instruction, this function stores the line and column of the executing instruction to a coverage file.
+Study the implementation to understand the APIs used for injecting the function.
+You will implement a dynamic analysis that tracks the kind, location as well as the runtime values of the operands to a binary operator.
+For this you will have to check if an instruction is a `BinaryOperator` and instrument it with the `instrumentBinOpOperands` function, which you will be implementing next.
+The `instrumentBinOpOperands` function has to inject calls to `__binop_op__` before every binary operator.
+You can see that `__binop_op__` takes 5 arguments, namely, a symbol for the operator, the line and column of the operation and the runtime values of the two operands.
+You can use the `getBinOpSymbol` function to get the symbol corresponding to the operator.
+Inorder to get the runtime values of the operands, it is necessary to keep in mind that in LLVM **a variable defined by an instruction is represented by the instruction itself**.
+
+**_Code Coverage Primer_**.
+Code coverage is a measure of how much of a program’s code is executed in a particular run.
+There are a number of different criterias to describe coverage.
+In this lab we are providing line coverage and you are implementing an artificial criteria of tracking binary operators during the execution of a program using the same mechanisms underlying modern code coverage tools, such as the LLVM’s source-based code coverage tool and gcov.
+It instruments the program’s LLVM IR instructions at compile-time to record the line and column number of the program’s source-level instructions that are executed at run-time.
+This seemingly primitive information enables powerful software analysis use-cases.
+In the next lab, you will use line coverage information to guide an automated test input generator, thereby realizing the architecture of modern industrial-strength fuzzers.
+
+<img src="../images/lab2/code_coverages_example.png" style="height: auto; width: 100%">
```

Comment:
Change image name to `example-coverage-report.png`

Update the src url similar to what was mentioned in the comment above.